File: java/graph/graphTraversal.java
package graph;
import java.util.LinkedList;


public class graphTraversal {
    public static void bfs(int[][] adjMatrix) {
        int n = adjMatrix.length;
        boolean[] visited = new boolean[n];

        LinkedList<Integer> queue = new LinkedList<Integer>();

        int startNode = 0;
        visited[startNode] = true;
        queue.add(startNode);

        System.out.println("BFS Traversal:");

        while (queue.size() != 0) {

            int currentVertex = queue.poll();
            System.out.print(currentVertex + " ");

            for (int i = 0; i < n; i++) {
                if (adjMatrix[currentVertex][i] == 1 && !visited[i]) {
                    visited[i] = true;
                    queue.add(i);
                }
            }
        }
        System.out.println();
    }

    public static void dfs(int[][] adjMatrix) {
        int n = adjMatrix.length;
        boolean[] visited = new boolean[n];

        System.out.println("\nDFS Traversal:");

        dfsRecursive(0, adjMatrix, visited);
        System.out.println();
    }

    private static void dfsRecursive(int currentVertex, int[][] adjMatrix, boolean[] visited) {

        visited[currentVertex] = true;
        System.out.print(currentVertex + " ");

        int n = adjMatrix.length;
        for (int i = 0; i < n; i++) {
            if (adjMatrix[currentVertex][i] == 1 && !visited[i]) {
                dfsRecursive(i, adjMatrix, visited);
            }
        }
    }

    public static void addEdge(int[][] adj, int s, int d) {
        if (s < 0 || s >= adj.length || d < 0 || d >= adj.length) {
            System.out.println("invalid edge");
            return;
        }
        adj[s][d] = 1;
        adj[d][s] = 1; // undirected graph
    }

    public static void main(String args[]) {
        int[][] adj = { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } };
        addEdge(adj, 0, 1);
        addEdge(adj, 0, 2);
        addEdge(adj, 0, 2);

        bfs(adj);
        dfs(adj);
    }
}

File: python/graphTraversal.py
graph = {
    'A': ['D'],
    'B': ['A', 'E', 'C'],
    'C': ['E'],
    'D': ['C', 'G'],
    'E': ['A'],
    'F': ['C'],
    'G': ['F', 'H'],
    'H': []
}

def bfs(graph, start):
    visited = []
    queue = [start]
    traversal = []

    while queue:
        vertex = queue.pop(0)  
        if vertex not in visited:
            visited.append(vertex)
            traversal.append(vertex)
            for n in graph[vertex]:
                if n not in visited and n not in queue:
                    queue.append(n)
    return traversal

def dfs(graph, start, visited=None, traversal=None):
    if visited==None:
        visited = []
    if traversal==None:
        traversal = []

    visited.append(start)
    traversal.append(start)

    for n in graph[start]:
        if n not in visited:
            dfs(graph, n, visited, traversal)

    return traversal

start = 'B'
print("BFS Traversal from", start, " is ", bfs(graph, start))
print("DFS Traversal from", start, " is ", dfs(graph, start))
